crmprtd/insert.py

```diff
@@ -54,6 +54,22 @@ def max_power_of_two(num):
     return 2 ** floor(mathlog(num, 2))
 
 
+def get_network_name(sesh, obs):
+    Obs_var = sesh.query(Variable).filter_by(id=obs.vars_id).first()
+    return Obs_var.network.name
+
+
+def obs_by_network(observations, sesh):
+    observations.sort(key=lambda obs: get_network_name(sesh, obs))
+    obs_by_network_dict = {
+        network_name: list(obs_group)
+        for network_name, obs_group in groupby(
```

Comment:
Have you considered the pros/cons of simply returning the return value of `groupby()` here? Could you discuss them?

Reply:
From the [docs](https://docs.python.org/3/library/itertools.html#itertools.groupby):
“The returned group is itself an iterator that shares the underlying iterable with groupby(). Because the source is shared, when the groupby() object is advanced, the previous group is no longer visible. So, if that data is needed later, it should be stored as a list”

Since we are planning to pass lists of observations grouped by their associated network, using dictionary comprehension with network as the key and a list of observations as the value is preferred.

The pros that come to mind are conciseness and memory efficiency. The return value would be a k,v pair that gets immediately consumed. This means that once the observations associated with a key have been iterated we don’t store them in memory as we do with the dictionary approach.